lib/backend/k8s/resources/globalbgppeer.go

```diff
@@ -0,0 +1,153 @@
+// limitations under the License.
+
+package resources
+
+import (
+	"reflect"
+	"strings"
+
+	"github.com/projectcalico/libcalico-go/lib/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/k8s/thirdparty"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/converter"
+	"github.com/projectcalico/libcalico-go/lib/net"
+	"github.com/projectcalico/libcalico-go/lib/scope"
+
+	log "github.com/Sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	GlobalBGPPeerResourceName = "globalbgppeers"
+	GlobalBGPPeerTPRName      = "global-bgp-peer.projectcalico.org"
+)
+
+func NewGlobalBGPPeerHandler(c *kubernetes.Clientset, r *rest.RESTClient) ResourceHandler {
+	return newTPRHandler(
+		c, r,
+		GlobalBGPPeerTPRName,
+		GlobalBGPPeerResourceName,
+		"Calico Global BGP Peers",
+		reflect.TypeOf(thirdparty.GlobalBGPPeer{}),
+		reflect.TypeOf(thirdparty.GlobalBGPPeerList{}),
+		GlobalBGPPeerConverter{},
+	)
+}
+
+// GlobalBGPPeerConvert implements the ResourceConverter interface.
+type GlobalBGPPeerConverter struct {
+	converter.BGPPeerConverter
+}
+
+func (_ GlobalBGPPeerConverter) ListInterfaceToKey(l model.ListInterface) model.Key {
+	pl := l.(model.GlobalBGPPeerListOptions)
+	if pl.PeerIP.IP != nil {
+		return model.GlobalBGPPeerKey{PeerIP: pl.PeerIP}
+	}
+	return nil
+}
+
+func (_ GlobalBGPPeerConverter) KeyToResourceName(k model.Key) (string, error) {
+	// The name should be policed before we get here.
+	pk := k.(model.GlobalBGPPeerKey)
+
+	// Convert the IP address to a k8s compatible name by replacing periods
+	// and colons with dashes.
+	return IPToResourceName(pk.PeerIP), nil
+}
+
+func (c GlobalBGPPeerConverter) ResourceToKVPair(r k8sResource) (*model.KVPair, error) {
+	// Since we are using the Calico API Spec definition to store the Calico
+	// BGP Peer, use the client conversion helper to convert between KV and API.
+	t := r.(*thirdparty.GlobalBGPPeer)
+	ip := ResourceNameToIP(t.Metadata.Name)
+	peer := api.BGPPeer{
+		Metadata: api.BGPPeerMetadata{
+			PeerIP: *ip,
+			Scope:  scope.Global,
+		},
+		Spec: t.Spec,
+	}
+	kvp, err := c.ConvertAPIToKVPair(peer)
+	if err != nil {
+		return nil, err
+	}
+	kvp.Revision = t.Metadata.ResourceVersion
+
+	return kvp, nil
+}
+
+func (c GlobalBGPPeerConverter) KVPairToResource(kvp *model.KVPair) (k8sResource, error) {
+	r, err := c.ConvertKVPairToAPI(kvp)
+	if err != nil {
+		return nil, err
+	}
+
+	tprName, err := c.KeyToResourceName(kvp.Key)
+	if err != nil {
+		return nil, err
+	}
+
+	tpr := thirdparty.GlobalBGPPeer{
+		Metadata: metav1.ObjectMeta{
+			Name: tprName,
+		},
+		Spec: r.(*api.BGPPeer).Spec,
+	}
+	if kvp.Revision != nil {
+		tpr.Metadata.ResourceVersion = kvp.Revision.(string)
+	}
+	return &tpr, nil
+}
+
+// ResourceNameToIP converts a name used for a k8s resource to an IP address.
+func ResourceNameToIP(name string) *net.IP {
```

Comment:
I *think* I should remove this method and just get the IP from the Spec and then ignore entries where Name and IP don't match.  WDYT?